### PR TITLE
[docs] Add scikit-learn for intersphinx (fixes #5954)

### LINKED ISSUE #5954: [ci] [docs] docs rendering failing with 'undefined label: 'metadata_routing'
## Description

The `check-docs` CI job is failing with this error that seems to be occurring while rendering the sphinx docs to HTML

> Warning, treated as error:
/home/runner/work/LightGBM/LightGBM/python-package/lightgbm/dask.py:docstring of lightgbm.dask.DaskLGBMClassifier:1:undefined label: 'metadata_routing'

([build link](https://github.com/microsoft/LightGBM/actions/runs/5424541292/jobs/9864047043?pr=5953))

## Reproducible example

I see that in multiple CI builds here, including the latest on `master` ([build link](https://github.com/microsoft/LightGBM/actions/runs/5423459086/jobs/9861417030)).

I'm able to reproduce this locally in a Python 3.10 environment on macOS, with the following:

```shell
mamba install --yes \
    breathe \
    doxygen \
    scikit-learn \
    sphinx \
    sphinx_rtd_theme \
    'scikit-learn>=1.3'
```

## Additional Notes

When downgrading from `scikit-learn` 1.3.0 to 1.2.2, this error goes away and the docs render correctly. I strongly suspect that it's related to changes in the most recent version of `scikit-learn`.

### PR DESCRIPTION
**Pull request by thomasjpfan**
Fixes https://github.com/microsoft/LightGBM/issues/5954

This PR enables Sphinx's intersphinx extension such that links to scikit-learn's docs can be resolved. Note that the [sphinx.ext.intersphinx](https://www.sphinx-doc.org/en/master/usage/extensions/intersphinx.html) extension comes with Sphinx.

### CODE REVIEW

**Review by jameslamb — APPROVED** [review]
This is great, thanks so much @thomasjpfan !

I think this is a much better solution than what I proposed in https://github.com/scikit-learn/scikit-learn/pull/26747. It's lightweight for any project to implement, and gives a much stronger guarantee of being future-proof against future changes upstream in `scikit-learn`.

Appreciate you taking the time to open this 😊
